cohorts/load.py

```diff
@@ -313,12 +313,42 @@ def load_from_cache(self, cache_name, patient_id, file_name):
         if not path.exists(cache_file):
             return None
 
+        df_provenance = self.generate_provenance(cache_name, patient_id)
+        df_provenance_previous = pd.read_csv(path.join(patient_cache_dir, "PROVENANCE"), index_col=None)
+
+        def provenance_set_from_df(df):
+            return set(dict(df.to_dict("split")["data"]).items())
+
+        def provenance_str_from_set(s):
+            return ["%s==%s" % (key, value) for (key, value) in s]
+
+        provenance_set = provenance_set_from_df(df_provenance)
+        provenance_set_previous = provenance_set_from_df(df_provenance_previous)
+        new_provenance_diff = provenance_set.difference(provenance_set_previous)
+        old_provenance_diff = provenance_set_previous.difference(provenance_set)
+        if new_provenance_diff != set():
+            print("Current provenance for patient %s: %s" % (
+                patient_id, provenance_str_from_set(
+                    new_provenance_diff)))
+        if old_provenance_diff != set():
+            print("Cached provenance for patient %s: %s" % (
+                patient_id, provenance_str_from_set(
+                    old_provenance_diff)))
+
         if path.splitext(cache_file)[1] == ".csv":
             return pd.read_csv(cache_file, dtype={"patient_id": object})
         else:
             with open(cache_file, "rb") as f:
                 return pickle.load(f)
 
+    def generate_provenance(self, cache_name, patient_id):
+        df_provenance = pd.DataFrame()
```

Comment:
is this a data frame just because it is easier serialize/deserialize from the disk? 
Out of curiosity: are `Dict`/`JSON`s harder to save/load from the disk in Python?